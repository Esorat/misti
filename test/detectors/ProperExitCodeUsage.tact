// Bad example - using reserved exit code 128, which should trigger the detector warning
contract BadExitCodeContract {
    const InvalidExitCode: Int = 128;
    owner: Address;

    init(ownerAddress: Address) {
        self.owner = ownerAddress;
    }

    receive("Bad Example!") {
        nativeThrowUnless(self.InvalidExitCode, sender() == self.owner);
    }
}

// Another bad example - using an exit code above the allowed range, which should also trigger a warning
contract ExcessiveExitCodeContract {
    const ExcessiveExitCode: Int = 70000;
    owner: Address;

    init(ownerAddress: Address) {
        self.owner = ownerAddress;
    }

    receive("Bad Example!") {
        nativeThrowUnless(self.ExcessiveExitCode, sender() == self.owner);
    }
}

// Good example - using a valid developer-defined exit code within the allowed range
contract ValidExitCodeContract {
    const ValidExitCode: Int = 256;
    owner: Address;

    init(ownerAddress: Address) {
        self.owner = ownerAddress;
    }

    receive("Good example") {
        nativeThrowUnless(self.ValidExitCode, sender() == self.owner);
    }
}

// Another good example - an exit code within the range, near the upper limit
contract UpperLimitExitCodeContract {
    const NearMaxExitCode: Int = 65535;
    owner: Address;

    init(ownerAddress: Address) {
        self.owner = ownerAddress;
    }

    receive("upperLimitFunc") {
        nativeThrowUnless(self.NearMaxExitCode, sender() == self.owner);
    }
}