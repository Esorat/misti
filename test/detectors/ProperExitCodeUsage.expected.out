[HIGH] ProperExitCodeUsage: Improper exit code used: 128.
test/detectors/ProperExitCodeUsage.tact:11:9:
  10 |     receive("Bad Example!") {
> 11 |         nativeThrowUnless(self.InvalidExitCode, sender() == self.owner);
               ^
  12 |     }
Use exit codes within the allowed range (256-65535, excluding 128).
Help: Adjust exit codes accordingly.
See: https://nowarp.io/tools/misti/docs/detectors/ProperExitCodeUsage

[HIGH] ProperExitCodeUsage: Improper exit code used: 70000.
test/detectors/ProperExitCodeUsage.tact:25:9:
  24 |     receive("Bad Example!") {
> 25 |         nativeThrowUnless(self.ExcessiveExitCode, sender() == self.owner);
               ^
  26 |     }
Use exit codes within the allowed range (256-65535, excluding 128).
Help: Adjust exit codes accordingly.
See: https://nowarp.io/tools/misti/docs/detectors/ProperExitCodeUsage